otelconf/internal/provider/envprovider_test.go

```diff
@@ -0,0 +1,143 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package provider // import "go.opentelemetry.io/contrib/internal/provider"
+
+import (
+	"errors"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestReplaceEnvVar(t *testing.T) {
+	for _, tt := range []struct {
+		name       string
+		envVarName string
+		uri        string
+		val        string
+		wantValue  any
+		wantErr    error
+	}{
+		{
+			name:       "int value",
+			envVarName: "INT_VALUE",
+			uri:        "INT_VALUE",
+			val:        "1",
+			wantValue:  []byte("1"),
+		},
+		{
+			name:       "string value",
+			envVarName: "STRING_VALUE",
+			uri:        "STRING_VALUE",
+			val:        "this is a string",
+			wantValue:  []byte("this is a string"),
+		},
+		{
+			name:      "invalid env var name",
+			uri:       "$%&(*&)",
+			wantValue: []byte(nil),
+			wantErr:   errors.New("invalid environment variable name: $%&(*&)"),
+		},
+		{
+			name:      "unset value has default",
+			uri:       "THIS_VALUE_IS_NOT_SET:-this is a default value",
+			wantValue: []byte("this is a default value"),
+		},
+		{
+			name:      "unset value no default",
+			uri:       "THIS_VALUE_IS_NOT_SET",
+			wantValue: []byte(nil),
+		},
+		{
+			name:       "invalid variable type map",
+			uri:        "MAP_VALUE",
+			envVarName: "MAP_VALUE",
+			val: `key:
+  value: something`,
+			wantValue: []byte(nil),
+			wantErr:   errors.New("invalid value type: unsupported type=map[string]interface {} for retrieved config, ensure that values are wrapped in quotes"),
+		},
+		{
+			name:       "invalid variable type list",
+			uri:        "LIST_VALUE",
+			envVarName: "LIST_VALUE",
+			val:        `["one", "two"]`,
+			wantValue:  []byte(nil),
+			wantErr:    errors.New("invalid value type: unsupported type=[]interface {} for retrieved config, ensure that values are wrapped in quotes"),
+		},
+		{
+			name:       "handle invalid yaml",
+			uri:        "NOT_A_NUMBER_VALUE",
+			envVarName: "NOT_A_NUMBER_VALUE",
+			val:        `!!int NaN`,
+			wantValue:  []byte(nil),
+			wantErr:    errors.New("invalid value type: yaml: cannot decode !!str `NaN` as a !!int"),
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			if len(tt.envVarName) > 0 {
+				t.Setenv(tt.envVarName, tt.val)
+			}
+			got, err := replaceEnvVar(tt.uri)
+			require.Equal(t, tt.wantValue, got)
+			if tt.wantErr != nil {
+				require.ErrorContains(t, err, tt.wantErr.Error())
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestReplaceEnvVars(t *testing.T) {
```

Comment:
Can we add all test cases from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/data-model.md#environment-variable-substitution (the ones that are in the table)?

Reply:
Added a bunch more tests, still need to ensure all the scenarios are covered and passing, but it's underway

Reply:
There are 4 test cases missing, would it be ok to separate those somewhat edge cases into a followup PR? The env substitution support as is in this PR would be really useful